programs/fundamentals/matrix.sc

```diff
@@ -0,0 +1,176 @@
+    check_matrix(a);
+    check_matrix(b);
+
+    if(length(a)!=length(b) || length(a:0)!=length(b:0),
+        exit(print('Cannot add uneven matrices'))
+    );
+
+    sum = [];
+
+    loop(length(a),
+        width = _;
+        sum:width=[];
+        loop(length(a:0),
+            sum:width:_ = a:width:_ + b:width:_
+        )
+    );
+    sum
+);
+
+matrix_subtract(a, b) ->( //returns a new matrix whose values are the difference of the two input matrices (i.e a - b)
+    check_matrix(a);
+    check_matrix(b);
+
+    if(length(a)!=length(b) || length(a:0)!=length(b:0),
+        exit(print('Cannot subtract uneven matrices'))
+    );
+
+    sum = [];
+
+    loop(length(a),
+        width = _;
+        sum:width=[];
+        loop(length(a:0),
+            sum:width:_ = a:width:_ - b:width:_
+        )
+    );
+    sum
+);
+
+//This is complicated, because you can do many things here, depending if input is scalar, vector or matrix:
+//  scalar * scalar = error (y doin that op here)
+//  scalar * vector = error (y doin that op here)
+//  scalar * matrix = error (cannot do it in this order). Still gotta handle.
+//  vector * scalar = error (y doin that op here)
+//  vector * vector = matrix with rows same as length of first vector, columns same as that of second.
+//  vector * matrix = error unless vector length is equal to matrix rows (so treating vector as 1*m matrix, where m is rows of the matrix with which ur multiplying), in which case it's a vector
+//  matrix * scalar = matrix scaled by scalar
+//  matrix * vector = error unless vector length is equal to matrix columns (so treating vector as 1*n matrix, where n is columns of the matrix with which ur multiplying), in which case it's a vector
+//  matrix * matrix = regular matrix operation.
+
+matrix_multiply(a, b) ->(
+    if(type(a)=='list', //if not a list, must be a scalar
+        result = [];
+        if(is_vector(a), // vector * something
+            if(is_vector(b),//vector * vector = matrix
+                loop(length(a),
+                    width = _;
+                    result:_=[];
+                    loop(length(b),
+                        result:width:_ = a:width * b:_
+                    )
+                ),
+                is_matrix(b),//vector * matrix = vector
+                if(length(a)==length(b),
+                    loop(length(b),
+                        row=_;
+                        result:row = reduce(b:row, _a + (_ * a:row),0);
+                    ),
+                    exit(print(str('Cannot multiply vector %s by matrix %s in this order, they have mismatched sizes', a, b)))
+                ),
+                exit(print('Why are you doing that operation here?'))
+            ),//in case they inputted a list which wasn't a vector.
+            is_matrix(a), //matrix * something
+            if(is_vector(b), //matrix * vector = vector
+                
+            ),
```

Comment:
mat×mat and mat×vec yet to be implemented?

Reply:
Yes, I hadn't gotten round to it cos I kinda forogt about this until I saw ur other pr